File: digi_lib_app/ios/Runner/NativeRenderingPlugin.swift
import Flutter
import UIKit
import Foundation

/**
 * Native rendering plugin for iOS platform
 * 
 * This plugin provides document rendering capabilities through platform channels.
 * It serves as a fallback when FFI is not available.
 */
public class NativeRenderingPlugin: NSObject, FlutterPlugin {
    private static let channelName = "digi_lib_native_rendering"
    private var channel: FlutterMethodChannel?
    
    // Performance metrics
    private var renderCount = 0
    private var totalRenderTime: TimeInterval = 0
    private var errorCount = 0
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
        let instance = NativeRenderingPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        print("Native rendering plugin registered for iOS")
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isAvailable":
            result(isNativeRenderingAvailable())
        case "renderPage":
            handleRenderPage(call, result: result)
        case "extractText":
            handleExtractText(call, result: result)
        case "getPageCount":
            handleGetPageCount(call, result: result)
        case "getPerformanceMetrics":
            handleGetPerformanceMetrics(result: result)
        case "clearCache":
            handleClearCache(result: result)
        case "setLogLevel":
            handleSetLogLevel(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func isNativeRenderingAvailable() -> Bool {
        // Check if native rendering capabilities are available
        // This is a placeholder - in a real implementation, you would check for
        // PDF rendering frameworks like PDFKit, etc.
        return true // Placeholder - assume available for now
    }
    
    private func handleRenderPage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let filePath = args["filePath"] as? String,
              let page = args["page"] as? Int,
              let dpi = args["dpi"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
            return
        }
        
        // Execute rendering on background queue
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            do {
                let startTime = CFAbsoluteTimeGetCurrent()
                let imageData = try self?.renderPageNative(filePath: filePath, page: page, dpi: dpi)
                let endTime = CFAbsoluteTimeGetCurrent()
                
                self?.renderCount += 1
                self?.totalRenderTime += (endTime - startTime)
                
                let response: [String: Any] = [
                    "success": true,
                    "imageData": imageData as Any
                ]
                
                DispatchQueue.main.async {
                    result(response)
                }
            } catch {
                self?.errorCount += 1
                print("Failed to render page: \(error.localizedDescription)")
                
                let response: [String: Any] = [
                    "success": false,
                    "error": error.localizedDescription
                ]
                
                DispatchQueue.main.async {
                    result(response)
                }
            }
        }
    }
    
    private func handleExtractText(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let filePath = args["filePath"] as? String,
              let page = args["page"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
            return
        }
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            do {
                let text = try self?.extractTextNative(filePath: filePath, page: page)
                
                let response: [String: Any] = [
                    "success": true,
                    "text": text as Any
                ]
                
                DispatchQueue.main.async {
                    result(response)
                }
            } catch {
                self?.errorCount += 1
                print("Failed to extract text: \(error.localizedDescription)")
                
                let response: [String: Any] = [
                    "success": false,
                    "error": error.localizedDescription
                ]
                
                DispatchQueue.main.async {
                    result(response)
                }
            }
        }
    }
    
    private func handleGetPageCount(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let filePath = args["filePath"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing required arguments", details: nil))
            return
        }
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            do {
                let pageCount = try self?.getPageCountNative(filePath: filePath)
                
                let response: [String: Any] = [
                    "success": true,
                    "pageCount": pageCount as Any
                ]
                
                DispatchQueue.main.async {
                    result(response)
                }
            } catch {
                self?.errorCount += 1
                print("Failed to get page count: \(error.localizedDescription)")
                
                let response: [String: Any] = [
                    "success": false,
                    "error": error.localizedDescription
                ]
                
                DispatchQueue.main.async {
                    result(response)
                }
            }
        }
    }
    
    private func handleGetPerformanceMetrics(result: @escaping FlutterResult) {
        let metrics: [String: Any] = [
            "renderCount": renderCount,
            "totalRenderTime": totalRenderTime * 1000, // Convert to milliseconds
            "averageRenderTime": renderCount > 0 ? (totalRenderTime * 1000) / Double(renderCount) : 0,
            "errorCount": errorCount,
            "platform": "ios"
        ]
        result(metrics)
    }
    
    private func handleClearCache(result: @escaping FlutterResult) {
        // Clear any cached data
        // This is a placeholder - implement actual cache clearing logic
        print("Cache cleared")
        result(nil)
    }
    
    private func handleSetLogLevel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let args = call.arguments as? [String: Any],
           let level = args["level"] as? String {
            print("Log level set to: \(level)")
            // Implement log level setting logic
        }
        result(nil)
    }
    
    /**
     * Native rendering implementation
     * This is a placeholder implementation - in a real app, this would use
     * native PDF rendering frameworks like PDFKit, Core Graphics, etc.
     */
    private func renderPageNative(filePath: String, page: Int, dpi: Int) throws -> Data {
        // Validate file exists
        guard FileManager.default.fileExists(atPath: filePath) else {
            throw NSError(domain: "NativeRenderingError", code: 1, userInfo: [NSLocalizedDescriptionKey: "File not found: \(filePath)"])
        }
        
        // Validate parameters
        guard page >= 0 else {
            throw NSError(domain: "NativeRenderingError", code: 2, userInfo: [NSLocalizedDescriptionKey: "Page number must be non-negative"])
        }
        
        guard dpi > 0 && dpi <= 600 else {
            throw NSError(domain: "NativeRenderingError", code: 3, userInfo: [NSLocalizedDescriptionKey: "DPI must be between 1 and 600"])
        }
        
        // This is a mock implementation that returns a simple placeholder image
        // In a real implementation, you would:
        // 1. Load the document using PDFKit or Core Graphics
        // 2. Render the specified page at the given DPI
        // 3. Convert to WebP or PNG format
        // 4. Return the image data
        
        return generatePlaceholderImage(dpi: dpi)
    }
    
    private func extractTextNative(filePath: String, page: Int) throws -> String {
        guard FileManager.default.fileExists(atPath: filePath) else {
            throw NSError(domain: "NativeRenderingError", code: 1, userInfo: [NSLocalizedDescriptionKey: "File not found: \(filePath)"])
        }
        
        guard page >= 0 else {
            throw NSError(domain: "NativeRenderingError", code: 2, userInfo: [NSLocalizedDescriptionKey: "Page number must be non-negative"])
        }
        
        // Mock text extraction - in a real implementation, you would:
        // 1. Load the document using PDFKit
        // 2. Extract text from the specified page
        // 3. Return the text content
        
        return """
        Mock text content for page \(page + 1) from iOS platform channel.
        
        This is placeholder text that would be extracted from the actual document. 
        In a real implementation, this would use native PDF frameworks like PDFKit 
        to extract the actual text content from the document.
        """
    }
    
    private func getPageCountNative(filePath: String) throws -> Int {
        guard FileManager.default.fileExists(atPath: filePath) else {
            throw NSError(domain: "NativeRenderingError", code: 1, userInfo: [NSLocalizedDescriptionKey: "File not found: \(filePath)"])
        }
        
        // Mock page count - in a real implementation, you would:
        // 1. Load the document using PDFKit
        // 2. Get the actual page count
        // 3. Return the count
        
        // Return a mock page count based on file size
        do {
            let attributes = try FileManager.default.attributesOfItem(atPath: filePath)
            let fileSize = attributes[.size] as? Int64 ?? 0
            
            switch fileSize {
            case 0..<(1024 * 1024):
                return 5      // < 1MB
            case (1024 * 1024)..<(10 * 1024 * 1024):
                return 25     // < 10MB
            default:
                return 100
            }
        } catch {
            return 10 // Default fallback
        }
    }
    
    private func generatePlaceholderImage(dpi: Int) -> Data {
        // Generate a simple placeholder image
        // This is a minimal WebP-like header followed by some data
        let width = Int(8.5 * Double(dpi)) // 8.5 inch width
        let height = Int(11 * Double(dpi)) // 11 inch height (letter size)
        
        // Simple placeholder data - not a real WebP but serves for testing
        var headerData = Data([
            0x52, 0x49, 0x46, 0x46, // "RIFF"
            0x00, 0x00, 0x00, 0x00, // File size (placeholder)
            0x57, 0x45, 0x42, 0x50, // "WEBP"
            0x56, 0x50, 0x38, 0x20  // "VP8 "
        ])
        
        // Add some mock image data
        let mockData = Data((0..<100).map { UInt8($0 % 256) })
        headerData.append(mockData)
        
        return headerData
    }
}

File: digi_lib_app/android/app/src/main/kotlin/com/example/digi_lib_app/NativeRenderingPlugin.kt
package com.example.digi_lib_app

import android.content.Context
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File
import java.io.FileInputStream
import java.io.ByteArrayOutputStream
import java.util.concurrent.Executors
import java.util.concurrent.ExecutorService

/**
 * Native rendering plugin for Android platform
 * 
 * This plugin provides document rendering capabilities through platform channels.
 * It serves as a fallback when FFI is not available.
 */
class NativeRenderingPlugin : FlutterPlugin, MethodCallHandler {
    companion object {
        private const val CHANNEL = "digi_lib_native_rendering"
        private const val TAG = "NativeRenderingPlugin"
    }

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val executor: ExecutorService = Executors.newCachedThreadPool()
    
    // Performance metrics
    private var renderCount = 0
    private var totalRenderTime = 0L
    private var errorCount = 0

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        Log.d(TAG, "Native rendering plugin attached")
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        executor.shutdown()
        Log.d(TAG, "Native rendering plugin detached")
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "isAvailable" -> {
                result.success(isNativeRenderingAvailable())
            }
            "renderPage" -> {
                handleRenderPage(call, result)
            }
            "extractText" -> {
                handleExtractText(call, result)
            }
            "getPageCount" -> {
                handleGetPageCount(call, result)
            }
            "getPerformanceMetrics" -> {
                handleGetPerformanceMetrics(result)
            }
            "clearCache" -> {
                handleClearCache(result)
            }
            "setLogLevel" -> {
                handleSetLogLevel(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun isNativeRenderingAvailable(): Boolean {
        // Check if native rendering libraries are available
        // This is a placeholder - in a real implementation, you would check for
        // PDF rendering libraries like PDFium, MuPDF, etc.
        return try {
            // Try to load native library or check for system capabilities
            true // Placeholder - assume available for now
        } catch (e: Exception) {
            Log.w(TAG, "Native rendering not available: ${e.message}")
            false
        }
    }

    private fun handleRenderPage(call: MethodCall, result: Result) {
        val filePath = call.argument<String>("filePath")
        val page = call.argument<Int>("page")
        val dpi = call.argument<Int>("dpi")

        if (filePath == null || page == null || dpi == null) {
            result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
            return
        }

        // Execute rendering on background thread
        executor.execute {
            try {
                val startTime = System.currentTimeMillis()
                val imageData = renderPageNative(filePath, page, dpi)
                val endTime = System.currentTimeMillis()
                
                renderCount++
                totalRenderTime += (endTime - startTime)
                
                val response = mapOf(
                    "success" to true,
                    "imageData" to imageData
                )
                
                // Return result on main thread
                android.os.Handler(android.os.Looper.getMainLooper()).post {
                    result.success(response)
                }
            } catch (e: Exception) {
                errorCount++
                Log.e(TAG, "Failed to render page: ${e.message}", e)
                
                val response = mapOf(
                    "success" to false,
                    "error" to e.message
                )
                
                android.os.Handler(android.os.Looper.getMainLooper()).post {
                    result.success(response)
                }
            }
        }
    }

    private fun handleExtractText(call: MethodCall, result: Result) {
        val filePath = call.argument<String>("filePath")
        val page = call.argument<Int>("page")

        if (filePath == null || page == null) {
            result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
            return
        }

        executor.execute {
            try {
                val text = extractTextNative(filePath, page)
                
                val response = mapOf(
                    "success" to true,
                    "text" to text
                )
                
                android.os.Handler(android.os.Looper.getMainLooper()).post {
                    result.success(response)
                }
            } catch (e: Exception) {
                errorCount++
                Log.e(TAG, "Failed to extract text: ${e.message}", e)
                
                val response = mapOf(
                    "success" to false,
                    "error" to e.message
                )
                
                android.os.Handler(android.os.Looper.getMainLooper()).post {
                    result.success(response)
                }
            }
        }
    }

    private fun handleGetPageCount(call: MethodCall, result: Result) {
        val filePath = call.argument<String>("filePath")

        if (filePath == null) {
            result.error("INVALID_ARGUMENTS", "Missing required arguments", null)
            return
        }

        executor.execute {
            try {
                val pageCount = getPageCountNative(filePath)
                
                val response = mapOf(
                    "success" to true,
                    "pageCount" to pageCount
                )
                
                android.os.Handler(android.os.Looper.getMainLooper()).post {
                    result.success(response)
                }
            } catch (e: Exception) {
                errorCount++
                Log.e(TAG, "Failed to get page count: ${e.message}", e)
                
                val response = mapOf(
                    "success" to false,
                    "error" to e.message
                )
                
                android.os.Handler(android.os.Looper.getMainLooper()).post {
                    result.success(response)
                }
            }
        }
    }

    private fun handleGetPerformanceMetrics(result: Result) {
        val metrics = mapOf(
            "renderCount" to renderCount,
            "totalRenderTime" to totalRenderTime,
            "averageRenderTime" to if (renderCount > 0) totalRenderTime / renderCount else 0,
            "errorCount" to errorCount,
            "platform" to "android"
        )
        result.success(metrics)
    }

    private fun handleClearCache(result: Result) {
        try {
            // Clear any cached data
            // This is a placeholder - implement actual cache clearing logic
            Log.d(TAG, "Cache cleared")
            result.success(null)
        } catch (e: Exception) {
            result.error("CACHE_CLEAR_FAILED", e.message, null)
        }
    }

    private fun handleSetLogLevel(call: MethodCall, result: Result) {
        val level = call.argument<String>("level")
        if (level != null) {
            Log.d(TAG, "Log level set to: $level")
            // Implement log level setting logic
        }
        result.success(null)
    }

    /**
     * Native rendering implementation
     * This is a placeholder implementation - in a real app, this would use
     * native PDF rendering libraries like PDFium, MuPDF, etc.
     */
    private fun renderPageNative(filePath: String, page: Int, dpi: Int): ByteArray {
        // Validate file exists
        val file = File(filePath)
        if (!file.exists()) {
            throw IllegalArgumentException("File not found: $filePath")
        }

        // Validate parameters
        if (page < 0) {
            throw IllegalArgumentException("Page number must be non-negative")
        }
        if (dpi <= 0 || dpi > 600) {
            throw IllegalArgumentException("DPI must be between 1 and 600")
        }

        // This is a mock implementation that returns a simple placeholder image
        // In a real implementation, you would:
        // 1. Load the document using a PDF library
        // 2. Render the specified page at the given DPI
        // 3. Convert to WebP or PNG format
        // 4. Return the image bytes
        
        return generatePlaceholderImage(dpi)
    }

    private fun extractTextNative(filePath: String, page: Int): String {
        val file = File(filePath)
        if (!file.exists()) {
            throw IllegalArgumentException("File not found: $filePath")
        }

        if (page < 0) {
            throw IllegalArgumentException("Page number must be non-negative")
        }

        // Mock text extraction - in a real implementation, you would:
        // 1. Load the document using a PDF library
        // 2. Extract text from the specified page
        // 3. Return the text content
        
        return "Mock text content for page ${page + 1} from Android platform channel.\n\n" +
               "This is placeholder text that would be extracted from the actual document. " +
               "In a real implementation, this would use native PDF libraries to extract " +
               "the actual text content from the document."
    }

    private fun getPageCountNative(filePath: String): Int {
        val file = File(filePath)
        if (!file.exists()) {
            throw IllegalArgumentException("File not found: $filePath")
        }

        // Mock page count - in a real implementation, you would:
        // 1. Load the document using a PDF library
        // 2. Get the actual page count
        // 3. Return the count
        
        // Return a mock page count based on file size
        val fileSize = file.length()
        return when {
            fileSize < 1024 * 1024 -> 5      // < 1MB
            fileSize < 10 * 1024 * 1024 -> 25 // < 10MB
            else -> 100
        }
    }

    private fun generatePlaceholderImage(dpi: Int): ByteArray {
        // Generate a simple placeholder image
        // This is a minimal WebP-like header followed by some data
        val width = (8.5 * dpi).toInt() // 8.5 inch width
        val height = (11 * dpi).toInt() // 11 inch height (letter size)
        
        // Simple placeholder data - not a real WebP but serves for testing
        val headerData = byteArrayOf(
            0x52.toByte(), 0x49.toByte(), 0x46.toByte(), 0x46.toByte(), // "RIFF"
            0x00.toByte(), 0x00.toByte(), 0x00.toByte(), 0x00.toByte(), // File size (placeholder)
            0x57.toByte(), 0x45.toByte(), 0x42.toByte(), 0x50.toByte(), // "WEBP"
            0x56.toByte(), 0x50.toByte(), 0x38.toByte(), 0x20.toByte()  // "VP8 "
        )
        
        // Add some mock image data
        val mockData = ByteArray(100) { (it % 256).toByte() }
        
        return headerData + mockData
    }
}